nh.c

```diff
@@ -64,42 +64,43 @@ __attribute__((visibility("hidden"))) void nh_failsafe_destroy(nh_failsafe_t *fs
 // used afterwards.
 __attribute__((visibility("hidden"))) void nh_failsafe_uninstall(nh_failsafe_t *fs);
 
-// The following blacklists help prevent any whoopsies when deleting files during uninstall
+// The following blacklist helps prevent any whoopsies when deleting files or directories during uninstall
 
-// These files are important for recovery purposes.
-static const char *delete_file_blacklist[] = {
-    "/bin/sh",
-    "/etc/init.d/rcS",
-    "/usr/local/Kobo/pickel",
-};
-
-// Let's not allow deleting any files in /bin, /sbin, or u-boot stuff.
-static const char *delete_prefix_blacklist[] = {
+// Let's not allow deleting any files in /bin, /sbin, /etc/init.d or u-boot stuff.
+static const char *delete_blacklist[] = {
     "/bin",
     "/sbin",
     "/etc/u-boot",
+    "/etc/init.d",
+    "/usr/local/Kobo/pickel",
 };
 
 // This is for the schmuck who forgot to terminate his array of paths
 // and accidentally deleted '/bin/sh' due to an out-of-bounds read.
-void nh_delete_path(const char *path) {
-    if (path) {
-        for (unsigned int i = 0; i < (sizeof(delete_file_blacklist) / sizeof(*delete_file_blacklist)); i++) {
-            if (!strcmp(delete_file_blacklist[i], path)) {
-                nh_log("(NickelHook) not deleting blacklisted file %s", path);
-                return;
-            }
-        }
-        for (unsigned int i = 0; i < (sizeof(delete_prefix_blacklist) / sizeof(*delete_prefix_blacklist)); i++) {
-            if (!strncmp(delete_prefix_blacklist[i], path, strlen(delete_prefix_blacklist[i]))) {
-                nh_log("(NickelHook) not deleting %s with blacklisted prefix %s", path, delete_prefix_blacklist[i]);
+void nh_delete_path(const char *path, bool is_dir) {
+    if (!path) {
+        nh_log("(NickelHook) no path supplied");
+        return;
+    }
+    const char *canonical_path;
+    // lets not play the guessing game with PATH_MAX
+    // note, this is from GNU, so _GNU_SOURCE is required
+    if ((canonical_path = canonicalize_file_name(path))) {
```

Comment:
Note, I'm not opposed to using a preallocated buffer of size `PATH_MAX` if that is what is preferred.

Reply:
That is preferred, as I'm trying to avoid unnecessary allocations in NH itself.

Reply:
Do you prefer a static buffer? Or local? If static, does it need to be thread local?

Reply:
I'd prefer a non-static local buffer for this.